### ISSUE TITLE
SpringData Neo4j 4 + remote database = TimeTree doesn't work?

### ISSUE THREAD

**Issue by Lokymil** [graphaware/neo4j-timetree#52]
Hi,
I'm not sure if it's the best place to ask that kind of question but I don't know where I can ask it to quickly have an answer.

For my project, I use Spring Data Neo4j 4 and a remote instance of Neo4j. My application is a Spring Boot Application running on port 8090 and my Neo4j database run on 8030 port (everything in localhost). Everything works fine and now I want to use a TimeTree to manage multiples events. I add the TimeTree and Framework jar from GraphAware to my database plugins folder.
Here is my SDN 4 configuration file :

    Neo4jConfig extends Neo4jConfiguration {

        @Value("${neo4j.url}")
        private String // "http://localhost:8030"

        @Bean
        public Neo4jServer neo4jServer() {
            return new RemoteServer(url);
        }

        @Override
        public SessionFactory getSessionFactory() {
            return new SessionFactory("com.zenika.client.mood");
       }

        @Bean
        public Session getSession() throws Exception {
            return super.getSession();
        }
    }

Events can be of different kind (vote, poke, ...) then I want to have CustomTimeTree (eg. : one for each type event). The problem is, for that kind of tree, I need a Node object but I can't find a way to have it with SDN 4. If I use a SingleTimeTree, I need a GraphDatabaseService object but this object seems to exist in SDN (3 & 4) only for embedded database. Moreover, if I try to create my own custom object extending Node, I need the GraphDatabaseService I can't have because of the remote configuration...

I tried to use my database as an embedded one by using the file URI of my database for my configuration. This way, the TimeTree works fine but other functions of my application can't be used. Indeed, to allow the TimeTree manager to add its nodes to my database I must shutdown the neo4j server, otherwise I got this error :
    `org.neo4j.kernel.StoreLockException: Store and its lock file has been locked by another process`
For this case, here is my configuration and the class creating the TimeTree :

```
 Neo4jConfig extends Neo4jConfiguration {

        @Value("${neo4j.url}")
        private String // "http://localhost:8030"

        @Bean
        public Neo4jServer neo4jServer() {
            return new RemoteServer(url);
        }

        @Override
        public SessionFactory getSessionFactory() {
            return new SessionFactory("com.zenika.client.mood");
         }

        @Bean
        public Session getSession() throws Exception {
            return super.getSession();
        }

        @Bean
        public GraphDatabaseService getGraphDatabaseService() {
            return new GraphDatabaseFactory().newEmbeddedDatabase("[path to neo4j root folder]/data/graph.db");
        }
    }
```

```
public class GATimeTreeService {

    public void test() {
        SingleTimeTree stt = new SingleTimeTree(graphDatabaseService);
        TimeInstantVO timeInstantVO = new TimeInstantVO(new Date().getTime(), Resolution.MONTH.name(), "UTC+01:00");
        TimeInstant timeInstant = TimeInstant.fromValueObject(timeInstantVO);
        stt.getOrCreateInstant(timeInstant);
    }

}
```

I pretty lost because everything I found was for embedded database and I have no idea how to use GraphAware TimeTree with a remote database... So I have three questions.

Did I miss something?
Is communication with remote database for TimeTree a feature not available yet? If yes, when will it be available?
How can I use TimeTree in my situation without create and manage it by myself?

**Comment by ikwattro** [comment]
Hi @Lokymil 

You can of course use the TimeTree plugin with a remote database. The best configuration for you with SDN4 would be to use `Automatic Event Attachments` as explained here : https://github.com/graphaware/neo4j-timetree#automatic-event-attachment

You would need however to have a different POJO for each type of event for SDN.

With the remote configuration, you will not need to do anything except the configuration setup.

Let me know if this help.

Christophe

**Comment by vince-bickers** [comment]
@Lokymil 

I think you may be getting confused regarding the GraphDatabaseService bean. For a remote instance, your code doesn't have direct access to GraphDatabaseService, so your bean is creating a completely separate database from the remote one. Your application is not talking to the embedded instance at all. 

As @ikwattro says, for the remote server, simply install the TimeTree module in the normal way. Your application should need no additional configuration.

If you want to run your application against an Embedded database, you can use SDN 4.1 / OGM 2.0.1 which provides a driver for an embedded instance. 

The documentation for this is [here](http://docs.spring.io/spring-data/neo4j/docs/current/reference/html/#_configuring_the_embedded_driver).

**Comment by vince-bickers** [comment]
Reading through this again I'm not sure I've understand the real issue here. 

Is it that you don't know how to load the nodes that are updated by the TT plugin back into your SDN application? 

Or is that you want to have events fired back into your application from the TT plugin?

Or something else entirely?

**Comment by bachmanm** [comment]
The main thing to understand:
- TimeTree runs in the same JVM as the Neo4j database
- SDN4 runs in the same JVM as your application

In a remote scenario, where your app and Neo4j are in two separate JVMs, the two can only interact via HTTP REST. Within that, the options are:
- REST APIs exposed by TimeTree (SDN knows nothing about those)
- Cypher generated by SDN - you would have to model your tree in Java classes and annotate them
- Cypher written by you, executed by SDN (using `@Query`)

What is it you're trying to achieve exactly?

**Comment by Lokymil** [comment]
Thanks for your answers! Now I got a better understanding of how this works.

@vince-bickers , @bachmanm For now, I am not able to create/use a TT in my remote database. What I'm trying to achieve is to create a TimeTree and use it link events to time with a specific resolution (probably month or day). Once those event or attached to TT, I will need to retrieve them according to the occurrence's time (eg. : retrieve event for April).
If I understand well, either I use the Automatic Event Attachments or I use the REST APIs exposed by the TT plugins to use TT in my remote database.

In the first case, the noe4j.properties described here has to be define :
https://github.com/graphaware/neo4j-timetree#automatic-event-attachment
Does this file has to be set in my application or in my database's plugins? According to the two separated JVM's it must be define in the TT plugin of my database, right?

In the second case, I have to call an URL like this (most probably POST) : 
http://my-server:my-port/graphaware/timetree/{rootNodeId}/single/{time as ms}?resolution=Month&timezone=GMT%2B1
With this URL, I will retrieve a node and I will have to manually link my event to this node or retrieve event linked to this node.

Knowing that I will have to manage multiple types of event with different configuration for each type (eg. : different resolution for each type). What is the best way to deal with my case? Using REST APIs seems to be easier.

**Comment by bachmanm** [comment]
Hi,

neo4j.properties is in the database config directory, indeed.

Regarding your second question, it depends on how many different types you have. Manual management using the REST API (or Cypher via stored procedures in soon to be released 3.0-compatible TimeTree) is definitely one option. The other option is to register multiple instances of the module, one for each kind of configuration.

**Comment by Lokymil** [comment]
@bachmanm Thank you for this explanation, I tried using the REST API and it works fine. I think i'll keep using this module this way.

Thank you all for your help!
